src/main/java/net/minecraftforge/registries/RegistryManager.java

```diff
@@ -133,9 +144,28 @@ public String getName()
             this.synced.add(name);
         for (ResourceLocation legacyName : builder.getLegacyNames())
             addLegacyName(legacyName, name);
+        if (registerToRoot && this == RegistryManager.ACTIVE && builder.getHasWrapper() && !Registry.REGISTRY.containsKey(reg.getRegistryName()) && Registry.REGISTRY instanceof WritableRegistry<?> rootRegistry)
+            registerRegistry(rootRegistry, reg);
         return getRegistry(name);
     }
 
+    @SuppressWarnings("unchecked")
+    private <T> void registerRegistry(WritableRegistry<T> rootRegistry, ForgeRegistry<?> forgeReg)
+    {
+        Registry<?> wrapper = forgeReg.getWrapper();
+        if (wrapper != null)
+            rootRegistry.register((ResourceKey<T>) forgeReg.getRegistryKey(), (T) wrapper, Lifecycle.experimental());
+    }
+
+    public static CompletableFuture<List<Throwable>> preNewRegistryEvent(final Executor executor,
+            final IModStateTransition.EventGenerator<? extends RegistryEvent.NewRegistry> eventGenerator)
+    {
+        return CompletableFuture.runAsync(() -> {
+            registerToRoot = true;
```

Comment:
This could use a brief comment explaining why it isn't just "always" true as it is not that obvious from looking at the code that it is because we don't want to register any of the registries that get staticly initialized in `GameData` to root.

Reply:
This is a good point. The real reason is that we don't want to try and register vanilla-wrapped registries twice. As Registry is patched with all the forge vanilla-wrapped registries which queries GameData, and then returns it back to vanilla to be registered to the root registry. So the RegistryManager will try to register all the GameData registries to the root registry first, and then the vanilla Registry patches will attempt it as well and crash. We could maybe instead simplify this by always registering to root and changing the patches in Registry to return the instance but *not* register it to the root registry. Thoughts?